lib/iris/analysis/stats.py

```diff
@@ -133,26 +70,47 @@ def pearsonr(cube_a, cube_b, corr_coords=None):
 
     """
 
-    # If no coords passed then set to all coords of cube.
+    # Assign larger cube to cube_1
+    if cube_b.ndim > cube_a.ndim:
+        cube_1 = cube_b
+        cube_2 = cube_a
+    else:
+        cube_1 = cube_a
+        cube_2 = cube_b
+
+    dim_coords_1 = [coord.name() for coord in cube_1.dim_coords]
+    dim_coords_2 = [coord.name() for coord in cube_2.dim_coords]
+    common_dim_coords = list(set(dim_coords_1) & set(dim_coords_2))
+    # If no coords passed then set to all common dimcoords of cubes.
     if corr_coords is None:
-        corr_coords = [c.name() for c in cube_a.dim_coords]
-
-    vec_a, vec_b, res_ind = _get_calc_view(cube_a,
-                                           cube_b,
-                                           corr_coords)
-
-    sa = vec_a - np.mean(vec_a, 0)
-    sb = vec_b - np.mean(vec_b, 0)
-    flat_corrs = np.sum((sa*sb), 0)/np.sqrt(np.sum(sa**2, 0)*np.sum(sb**2, 0))
-
-    corrs = flat_corrs.reshape([cube_a.shape[i] for i in res_ind])
-
-    # Construct cube to hold correlation results.
-    corrs_cube = iris.cube.Cube(corrs)
-    corrs_cube.long_name = "Pearson's r"
-    corrs_cube.units = "1"
-    for i, dim in enumerate(res_ind):
-        c = cube_a.dim_coords[dim]
-        corrs_cube.add_dim_coord(c, i)
-
-    return corrs_cube
+        corr_coords = common_dim_coords
+
+    # Broadcast weights to shape of cube_1 if necessary.
+    if weights is None or cube_1.shape == cube_2.shape:
+        weights_1 = weights
+        weights_2 = weights
+    else:
```

Comment:
Since we do a little messing around with the weights in the case of rebroadcast inputs, I think the responsibility for error handling might need to be taken here too. I think we just need to check that the shape of `weights_2` matches the shape of `cube_2` first before we do the `broadcast_to_shape` part. If it doesn't you can raise a `ValueError` explaining that the weights shape doesn't match the smaller of the inputs and is invalid. I can see people providing the weights for the larger cube by mistake and then getting weird error messages and I'd like to avoid that.